src/parser/mod.rs

```diff
@@ -103,6 +107,64 @@ pub struct KeyMetadata {
     pub size: u32,
 }
 
+/// A trait for parsing multiple types of values to memcached values.
+pub trait ToMemcachedValue {
+    /// Returns the length of the value in bytes.
+    fn length(&self) -> usize;
+    /// Writes the value to a writer.
+    fn write_to<W: AsyncWriteExt + Unpin>(
+        &self,
+        writer: &mut W,
+    ) -> impl Future<Output = Result<(), crate::Error>>;
+}
+
+impl ToMemcachedValue for &[u8] {
+    fn length(&self) -> usize {
+        <[u8]>::len(self)
+    }
+    async fn write_to<W: AsyncWriteExt + Unpin>(&self, writer: &mut W) -> Result<(), crate::Error> {
+        writer.write_all(self).await.map_err(crate::Error::from)
+    }
+}
+
+impl ToMemcachedValue for &str {
+    fn length(&self) -> usize {
+        <str>::len(self)
+    }
+    async fn write_to<W: AsyncWriteExt + Unpin>(&self, writer: &mut W) -> Result<(), crate::Error> {
+        writer
+            .write_all(self.as_bytes())
+            .await
+            .map_err(crate::Error::from)
+    }
+}
+
+macro_rules! impl_to_memcached_value_for_uint {
+    ($ty:ident) => {
+        impl ToMemcachedValue for $ty {
+            fn length(&self) -> usize {
+                // std::mem::size_of_val(self)??
```

Comment:
This just returns the size of the underlying value https://doc.rust-lang.org/std/mem/fn.size_of.html.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af98a07d34bca003d532b0cc711c4fe1

Reply:
Figured as much.  Copilot barfed that out at me and I left it there as a note to check later before the final cleanup pass.